Cut-point optimization for {BestBinaryNumericSplit} -- Issue[884]

nikolay-apanasov opened the pull request:
Added a cut-point optimization to {BestBinaryNumericSplit}, an optimization described by Fayyad, U.M. and Irani K.B. in "On the Handling of Continuous-Valued Attributes in Decision Tree Generation". Many more details can be found in the attached proposal for work on Issue[884].
[issue-884.pdf](https://github.com/mlpack/mlpack/files/14302229/issue-884.pdf)

shrit commented on src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp:
```diff
@@ -103,8 +103,24 @@ double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
       ++classCounts(sortedLabels[i], 1);
   }
 
+  // Flags for validating whether the current sample index k is
+  // eligible as a cut point. {repeat} if the value at index k is
+  // part of a repeating sequence. {class_changed} if the class at 
+  // index k changed from k-1. {eligible_cut} if the value did not 
+  // repeat and the class changed, meaning index k is a boundary 
+  // point. See Fayyad and Irani (1992) for more details on eligible 
+  // cut points.
```
Could you change the style of comment to be `/** */`

nikolay-apanasov replied:
Definitely: I just made a new commit. Does mlpack have a preferred workflow for pull requests?

nikolay-apanasov commented:
Any idea why the Python test suite failed? All the tests passed in my previous commit, and all that changed was the comment style in `BestBinaryNumericSplit`. Moreover the one test that I see failed was `hmm_train_test.cpp`: my fork has no modifications to the `hmm` package. I see that my fork is two commits behind `mlpack::master`, but at the time the test failed on 02.16, it was up to date.

rcurtin commented:
I think that it was just an unlucky test failure.  Some of the tests are probabilistic and while we do our best to keep the failure probability low enough that we don't see any failures, sometimes one will sneak through.  I'm not worried about it because your changes clearly don't affect the HMM code.  It may be worth more investigation on its own though.  (But again, outside the scope of this PR.)

nikolay-apanasov commented:
Ok, that makes sense.

rcurtin reviewed (commented):
Hey there @nikolay-apanasov, thank you so much for the contribution.  I really appreciate the well-written document you sent too.  The actual optimization makes sense to me and looks correct.  I think you are right that it will also apply to the Gini impurity and likely to the whole class of reasonable loss functions (but I did not do the exact math to check).

I'm concerned about what you wrote about the classification performance being different, though.  Ideally this is just an optimization that should not change the resulting trees at all.  Did you try running with the same random seed in your trials (if any randomness was even involved)?  I think that a regular decision tree is entirely deterministic, although a random forest is not.

I did a little benchmarking on a couple datasets.  Here's what I saw on my system just using `mlpack_decision_tree` with default arguments and `--minimum_leaf_size 1` (to build deeper trees):

* gyroscope (6 numeric, 3 categorical, 15M points)
  - 15.290s on `master`
  - 14.470s on this branch

* covertype (technically not fully numeric, but I treated one-hot features as numeric; 54 features, 581k points)
  - 18.462s on `master`
  - 17.146s on this branch

* mnist (784 numeric features, 60k points)
  - 5.799s on `master`
  - 5.622s on this branch
  
* higgs (29 numeric features, 11M points)
  - 1.995s on `master`
  - 1.977s on this branch (very minor speedup if any, probably it has to do with the data itself)

Nice acceleration!

If we can figure out the discrepancy in the accuracy measures, I think we're good to merge this.  I suspect that is just a matter of random seed or similar, because I can't see a reason why this would cause different results.

Also, do you want to add a note to `HISTORY.md`?  (And add your name to the list of authors, @mlpack-bot will say more about it when the PR is approved)

Just some notes on the other questions in the document you sent:
 * A `BestCategoricalBinarySplit` would be awesome.  The other ideas you proposed would be cool too.  It would be a good idea to benchmark them so that we can choose a reasonable default.
 * Linear combinations might work well, but my only concern is that the actual splitting process is enough of an inner loop that finding the linear combination may be sufficiently expensive that it would make training time go up noticeably.  (It may result in a better-performing tree, though.)
 * There's not currently a way to visualize a tree.  I'm not sure it's worth adding directly to mlpack, though, because generally decision trees are so large that any graphviz representation would be gigantic and probably unreadable anyway.  (Especially with mlpack where the focus is often on larger datasets.)

nikolay-apanasov commented:
Thank you @rcurtin  for the thoughtful and detailed review. 

> I'm concerned about what you wrote about the classification performance being different, though. Ideally this is just an optimization that should not change the resulting trees at all. Did you try running with the same random seed in your trials (if any randomness was even involved)? 

This got me thinking. I mentioned in the literature review that initially, I implemented this optimization incorrectly, because I forgot about a special case: when there is a sequence of samples with repeated values for a variable v_k, those samples can be rearranged in an equivalent sorted order. Therefore if there is such a sequence, and the samples in the sequence come from multiple classes, then the next cut point is a boundary point, independent of the classes of the samples on either side of the cut. And just now, I realized that I still did not get it right! The code handles the case of repeated values _preceding_ the cut point, but it does not check for repeats _following_ the cut point. Therefore, with enough sample points, the resulting trees could indeed differ.

Going over the code again, I also discovered one other problem. If the `DecisionTree` is run with `UseWeights`, I'm not sure that this optimization is theoretically sound. Fayyad and Irani do not address this variation in their paper. Until we can generalize their proof to this case, it would probably make sense to only run this optimization when `UseWeights` is false.

My apologies for the faulty code! I will address these issues and make another commit here soon. The second problem certainly was not what was causing the discrepancy in the accuracy, but I hope that the new logic that resolves the first issue will fix this. Yet another possibility is that Fayyad's proof does not generalize to the Gini index. In this case, given that there is a noticeable increase in performance for many datasets, if we run sufficient benchmarks and determine that, while there is a difference, for all practical purposes that difference is negligible, then in my opinion it would still be worth while to implement the optimization.

> Also, do you want to add a note to HISTORY.md? 

Ok, sounds good.

> A BestCategoricalBinarySplit would be awesome. The other ideas you proposed would be cool too. It would be a good idea to benchmark them so that we can choose a reasonable default ... Linear combinations might work well, but my only concern is that the actual splitting process is enough of an inner loop that finding the linear combination may be sufficiently expensive that it would make training time go up noticeably. (It may result in a better-performing tree, though.)

The proposed `KMeansBinaryCategoricalSplit` that finds locally optimal cut points via a clustering algorithm would be incredibly interesting to implement, and in some settings I imagine it would be just what you would want to use, but I don't think that it would be a reasonable default splitter. I could be totally wrong, though, and I agree that running benchmarks is the way to go. However I think it's going to be a while until I have the time to implement that splitter. After I implement `BestBinaryCategoricalSplit`, my thoughts were to propose an implementation of [Bayesian Additive Regression Trees](https://arxiv.org/pdf/0806.3286.pdf) for mlpack. Until then, `BestBinaryCategoricalSplit` is optimal, therefore it seems that this would be a logical default, except in the case when there are multiple classes and many categories.

Finally, when it comes to linear combinations, I agree with you completely. The description of this strategy in CART makes me think the same thing. Furthermore, in their commentary to Loh and Vanichsetakul (whose trees use linear combination splits), which was published in the same issue of ASA, Breiman and Friedman argued that 

> One general theme throughout the article is that splitting based on linear combinations of the variables is generally superior to univariate splits. Our experience has been to the contrary. Besides the obvious interpretability of models based on univariate splits (which is the biggest single advantage of the tree-structured approach) we have found that in most applications where recursive partitioning has higher accuracy than traditional methods, that advantage is achieved through univariate rather than linear combination splitting. Linear combination splitting was introduced in Friedman (1977). There, linear discriminant functions were applied recursively in a manner similar to that described in this article. This was later supplanted by the far more thorough algorithm implemented in CART. In both cases, the principal reason for introducing linear combination splits was to make recursive partitioning competitive with classical linear procedures in settings appropriate for the latter. It does this to some extent, but in these situations one is usually better off with the appropriate classical procedure. Although linear combination splitting has strong intuitive appeal, it does not seem to achieve this promise in practice except for the settings mentioned previously.

Given their indisputable expertise in the domain, reading that commentary does not make me eager to implement this split type! Especially given that it would take a substantial effort to get the implementation right. My main reason for including it in the review was for the sake of completeness, since it is indeed an interesting and unique variation.

> There's not currently a way to visualize a tree. I'm not sure it's worth adding directly to mlpack, though, because generally decision trees are so large that any graphviz representation would be gigantic and probably unreadable anyway. (Especially with mlpack where the focus is often on larger datasets.)

Ok. You are certainly right that this would not be useful for most people that create non-trivial trees. However, in my limited experience, it did help substantially in debugging, being able to compare trees on well designed test inputs. Hopefully my update will fix any discrepancies in the trees, and this won't be an issue.

nikolay-apanasov commented:
Yesterday in my free time, I played around with the implementation and refactored to properly account for the special case of repeat values before and after a potential cut point. No matter how I sliced it, the resulting code did not have any noticeable efficiency gains. Moreover, even in the case where the instantiation is `DecisionTree<InformationGain>`, that is when the tree is constructed with the purity function used in the proof by Fayyad and Irani, the resulting tree still differs by a very slight amount. It would be interesting to find out why the trees differ, given that the authors prove the correctness of the optimization. 

At this point it looks like this optimization might not be worth it after all, at least not the way that I implemented it. I will say that the authors of the optimization mention an alternative algorithm for computing the entropy while choosing a split, an algorithm that, unsurprisingly, only evaluates the entropy at the eligible cut points that fall between classes.  This algorithm is in section 4.2 of their paper, and it follows from Corollary 2 in that section. However implementing this algorithm would basically require completely re-implementing `SplitIfBetter`. Moreover, although they claim to achieve substantial efficiency gains, I am not inclined in this direction, because the description of the algorithm implicitly assumes that all sorted samples between eligible cut points all fall in the same class, strangely ignoring the special case in  their proof of _Theorem 1_  (which proves that all cut are boundary points),  that is the case of repeat values with multiple classes. Maybe you guys will have better luck, but to me it was not immediately obvious how you would extend their formulation in section 4.2 to handle this special case. It seems to me that my time would be much better spent implementing `BestBinaryCategoricalSplit`.

rcurtin commented:
At least personally, I think it would be useful to further investigate why the trees differ even when using information gain.  I wonder if the optimization is exposing some other problem.  Are you sure that when you tested the code, the same random seed and all other things were being held constant?  In your tests I don't think that you were using weights, so the optimization should still be valid when using the information gain.

(I also strongly suspect the theory is still valid for the Gini index and for weighted samples.  But given that we know it succeeds for the information gain, then we should see the same tree---unless the old code was actually incorrect!)

nikolay-apanasov commented:
@rcurtin It definitely is an interesting question, and I do suspect that there is a bug somewhere in the implementation, because a proof is a proof! Furthermore, I submitted a PR this weekend for `BestBinaryCategoricalSplit`, and I saw that you had worked on a very related issue in PR #3644. In fact, I was incredibly surprised by the analysis, and I did wonder again about what it was that was causing the discrepancy in the optimized trees.

My tests were indeed run with a consistent random seed. However I don't know what precisely you mean by "all other things were being held constant?".  The tests were being run with the same inputs to the `DecisionTree` constructor, with the same random seed. All that I was (explicitly) varying was the few lines in `BestBinaryNumericSplit` that related to the optimization. However, my experience over the years has led me to a place where I  assume that I never actually really know what is going on in a computer system, given that it runs on hardware that is designed and constructed by human beings.  My class on Digital Design and Integrated Circuits really helped solidified this point of view. For example, towards the end of the course we briefly covered  faults in integrated circuits, and in particular we covered [Single Event Effects](http://old.in2p3.fr/actions/formation/microelectronique11/Velazco%20%20SEU%20and%20Latchup.pdf). Bits do flip, even if the event is relatively unlikely.

rcurtin commented:
If you were running on reasonable data, I strongly doubt you encountered the issue in #3644.  That issue has gone undetected for years because the unspeakably vast majority of datasets don't happen to have values off by one bit in their floating point representations.

It is potentially true that a bit may have flipped during a run.  But the probability is so vanishingly small that I wouldn't even bother considering it.  More likely sources of bit flips in reality are power supply issues or RAM failure, and either of those are manifested by unstable systems that often reboot or crash.  Plus, in the extremely unlikely event that your hardware was fine *but* a freak event meant that a bit flipped, well, run it again and it'll be fine.  Or run it on a different computer and it'll be fine.  So I also suspect you did not encounter that.

In any case, personally speaking I don't have the time to figure out the exact issue of what is going on here, but if you are able to I would certainly appreciate it and welcome a fix if one is found.  I'd love to get the original optimization included one way or another. :+1:

nikolay-apanasov commented:
Given how simple the optimization is, and the fact that the trees are consistently different,  I also don't think that the problem is adjacent floating point values or SEE. I only mentioned SEE in passing because in my mind, it's an ever-present reminder to not exclude the incredible when analyzing a problem. 

It seems that this issue should not be inherently difficult to solve, but it may take a little while. I'd like to help out, certainly. The only thing is that I'm studying full-time right now, and so until summer, I can only find time for mlpack on the weekends. As I mentioned above, I decided I'm going to apply for GSOC: I'd like to add an implementation of Bayesian Additive Regression Trees to mlpack. Therefore the next few weeks I'm getting my proposal together. With that being said, I added this to my TODO list, and I plan on solving this problem the first couple of weekends in April. I'll reopen this PR if and when I get it solved.
